design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+
+TBD: Define a Java API method for variable substitution.
+
+In headless, the session is opened when starting the ksqlDB server with a SQL script. The variables stay alive until the server is
+killed.
+
+Variables are not persisted anywhere. They are living in memory as long as the session is not terminated.
+
+#### RUN SCRIPT command
+
+If a `RUN SCRIPT` command is executed in a CLI session, then any variable defined there will be local to the execution context of the
+script. The variables will override any variable previously defined in the session. When the script is completed, all local variables
+defined will be destroyed. This ensures users do not override their session variables by accident when running `RUN SCRIPT` commands.
+
+#### The SOURCE command
+
+To allow users load variables and keep their scope in the session, then the new `SOURCE` command will be added. This will override
+any previously defined variables, and will be destroyed until the user leaves the session.
+
+Syntax:
+```
+SOURCE '<script_file>';
+```
+
+The `SOURCE` command will work like a `RUN SCRIPT`. If other SQL statements are in the file, then those statements will also be
+executed, thus to avoid limiting users from using DDL/DML commands while defining variables at the same time from one single script.
+
+`SOURCE` will be used only on the CLI and headless mode. REST API will not accept it because files may be out of the scope of the server filesystem.
+
+### 3. Referencing substitution variables
+
+Variables will be referenced by wrapping the variable between `${}` characters (i.e. `${replicas}`).
+
+Noe: We need to make changes in QTT tests to use `${}` references instead of `{}` for consistency.
+
+Example:
+```
+ksql> DEFINE format = 'AVRO';
+ksql> DEFINE replicas = '3';
+ksql> CREATE STREAM stream1 (id INT) WITH (kafka_topic='stream1', value_format='${format}', replicas=${replicas});
+```
+
+Substitution will not attempt to add single-quotes to the values. Users will need to know the data type to use when using a variable. The ksqlDB
+server will be in charge of parsing and failing if the type is not allowed.
+
+Note: Variables are case-insensitive. A reference to `${replicas}` is the same as `${REPLICAS}`.
+
+### 4. Context for substitution variables
+
+Variable substitution will be allowed in specific SQL statements. They can be used to replace text and non-text literals, and identifiers such as
+column names and stream/table names. Variables cannot be used as reserved keywords.
+
+For instance:
+
+Note: `{VAR}` is used a reference where variables are permitted.
+```
+ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
+      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
+      
+ksql> INSERT INTO ${streamName} (${colName1}, ${colName2}) \
+      VALUES (${val1}, '${val2}');
```

Comment:
It's interesting to me that the contents of variables are substituted into the statement directly, and it's left to the user to get the quoting of the variables correct. This makes sense but it'd be great to have helpful error messages when this goes wrong. For example, if the user mistakenly submits
```
ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
      WITH (kafka_topic=${topicName}, format='${format}', replicas=${replicas}, ...);
```
instead of
```
ksql> CREATE STREAM ${streamName} (${colName1} INT, ${colName2} STRING) \
      WITH (kafka_topic='${topicName}', format='${format}', replicas=${replicas}, ...);
```
where the difference is that they're missing quotes around the topic name, it'd be great to have a useful error message so they aren't left guessing in the dark.

Reply:
Right. I'll try to add a customized error. I was thinking on leverage the parsing errors, which displays an error when using `${topicName}`. I'll see if the second parser can detect those undesired places.